### PR TITLE
rewrite vmaas errata endpoint in go

### PR DESCRIPTION
**Pull request by Dugowitch**
There is `StringSlice` type because, according to the spec, some values in the request can be either string or []string. After a conversation with @MichaelMraka, I implemented a custom JSON parser that will always produce []string. This seems to be a better approach as opposed to generics shenanigans or so.

### CODE REVIEW

**Review by psegedy — COMMENTED** [review]
Just some question, overall it looks good. I'll try to compare the response to python implementation

File: vmaas/types.go [hunk 1]
```diff
@@ -55,6 +55,69 @@ type CvesRequest struct {
...
+		res = append(res, val...)
+		res = append(res, ',')
+	}
+
+	res[len(res)-1] = ']' // replace the trailing comma with the closing ]
+	return res, nil
+}
+
+type ErrataRequest struct {
+	ErrataList    []string    `json:"errata_list"`
```
Comment (psegedy) [inline vmaas/types.go:112]: nitpick, we know it is a list
```suggestion
	Errata    []string    `json:"errata_list"`
```

File: vmaas/errata.go [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+	"github.com/pkg/errors"
+	"github.com/redhatinsights/vmaas-lib/vmaas/utils"
+)
+
+type ErrataDetails map[string]ErratumDetail
+
+type Errata struct {
+	ErrataList ErrataDetails `json:"errata_list"`
+	Type       []string      `json:"type"`
+	Severity   StringSlice   `json:"severity,omitempty"`
```
[the remark this answers is not in the document]
Comment (MichaelMraka) [reply vmaas/errata.go:15]: the reason for this is custom defined json marshal/unmarshal functions.
(severity can be a string or list of strings)

File: vmaas/errata.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+	"github.com/pkg/errors"
+	"github.com/redhatinsights/vmaas-lib/vmaas/utils"
+)
+
+type ErrataDetails map[string]ErratumDetail
+
+type Errata struct {
+	ErrataList ErrataDetails `json:"errata_list"`
+	Type       []string      `json:"type"`
```
Comment (psegedy) [inline vmaas/errata.go:14]: it's not in spec but the original implementation omits `Type` in the response when it is empty

File: vmaas/errata.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+}
+
+func (req *ErrataRequest) errata(c *Cache) (*Errata, error) { // TODO: implement opts
+	if len(req.ErrataList) == 0 {
+		return nil, errors.New("errata_list must contain at least one item")
+	}
+
+	errata := utils.TryExpandRegexPattern(req.ErrataList, c.ErratumDetails)
+	errata = filterInputErrata(c, errata, req)
+	errata, paginationDetails := utils.Paginate(errata, req.PageNumber, req.PageSize)
```
Comment (psegedy) [inline vmaas/errata.go:91]: errata slice needs to be sorted before pagination, we are getting different response for the same request e.g. with page_size=10

**Review by psegedy — CHANGES_REQUESTED** [review]

File: vmaas/errata.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	erratumRepos := c.ErratumID2RepoIDs[erratumID]
+	releaseVers := make([]string, 0, len(erratumRepos))
+	for repoID := range erratumRepos {
+		repoDetail := c.RepoDetails[repoID]
+		releaseVer := repoDetail.Releasever
+		if releaseVer != "" {
+			releaseVers = append(releaseVers, releaseVer)
+		}
+	}
+	return releaseVers
```
Comment (psegedy) [inline vmaas/errata.go:68]: releasever contains duplicates, filter them out please

**Review by psegedy — CHANGES_REQUESTED** [review]
we also need struct tags for types.Module struct and I found another issue that we have 2 StreamIDs for the same module streams, I'm not sure if that is an issue during the db export, in load or we just have to combine them into a single module in response (becasue they are really the same module but with different packages)

response from old python implementation:
```
"modules_list": [
        {
          "module_name": "idm",
          "module_stream": "DL1",
          "module_version": 8040020231123154610,
          "module_context": "5b01ab7e",
          "package_list": [
            "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.ppc64le",
            "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.ppc64le",
            "softhsm-devel-2.6.0-5.module+el8.4.0+10227+076cd560.ppc64le",
            "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.ppc64le",
            "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.ppc64le",
            "ipa-client-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
            "ipa-client-epn-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
            "ipa-client-samba-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
            "ipa-server-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
            "ipa-server-trust-ad-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
            "python3-qrcode-core-5.1-12.module+el8.1.0+4098+f286395e.noarch",
            "custodia-0.6.0-3.module+el8.1.0+4098+f286395e.noarch",
            "python3-custodia-0.6.0-3.module+el8.1.0+4098+f286395e.noarch",
            "python3-jwcrypto-0.5.0-1.module+el8.1.0+4098+f286395e.noarch",
            "python3-pyusb-1.0.0-9.module+el8.1.0+4098+f286395e.noarch",
            "python3-qrcode-5.1-12.module+el8.1.0+4098+f286395e.noarch",
            "python3-yubico-1.3.2-9.module+el8.1.0+4098+f286395e.noarch",
            "python3-kdcproxy-0.4-5.module+el8.2.0+4691+a05b2456.noarch",
            "ipa-healthcheck-0.7-3.module+el8.4.0+9007+5084bdd8.noarch",
            "ipa-healthcheck-core-0.7-3.module+el8.4.0+9007+5084bdd8.noarch",
            "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.x86_64",
            "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.x86_64",
            "softhsm-devel-2.6.0-5.module+el8.4.0+10227+076cd560.x86_64",
            "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.x86_64",
            "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.x86_64",
            "ipa-client-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
            "ipa-client-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-client-epn-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
            "ipa-client-samba-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
            "ipa-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-python-compat-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-selinux-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-server-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
            "ipa-server-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-server-dns-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "ipa-server-trust-ad-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
            "python3-ipaclient-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "python3-ipalib-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "python3-ipaserver-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
            "python3-ipatests-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch"
          ],
          "source_package_list": [
            "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.src",
            "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.src",
            "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.src",
            "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.src",
            "ipa-4.9.2-10.module+el8.4.0+20744+ffd02a6d.src",
            "pyusb-1.0.0-9.module+el8.1.0+4098+f286395e.src",
            "python-jwcrypto-0.5.0-1.module+el8.1.0+4098+f286395e.src",
            "custodia-0.6.0-3.module+el8.1.0+4098+f286395e.src",
            "python-qrcode-5.1-12.module+el8.1.0+4098+f286395e.src",
            "python-yubico-1.3.2-9.module+el8.1.0+4098+f286395e.src",
            "python-kdcproxy-0.4-5.module+el8.2.0+4691+a05b2456.src",
            "ipa-healthcheck-0.7-3.module+el8.4.0+9007+5084bdd8.src",
            "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.src",
            "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.src",
            "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.src",
            "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.src",
            "ipa-4.9.2-10.module+el8.4.0+20744+ffd02a6d.src"
          ]
        }
      ],
 ```
 
new response
```
            "modules_list": [
                {
                    "Name": "idm",
                    "StreamID": 87222,
                    "Stream": "DL1",
                    "Version": "8040020231123154610",
                    "Context": "5b01ab7e",
                    "PackageList": [
                        "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.ppc64le",
                        "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.ppc64le",
                        "softhsm-devel-2.6.0-5.module+el8.4.0+10227+076cd560.ppc64le",
                        "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.ppc64le",
                        "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.ppc64le",
                        "ipa-client-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
                        "ipa-client-epn-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
                        "ipa-client-samba-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
                        "ipa-server-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le",
                        "ipa-server-trust-ad-4.9.2-10.module+el8.4.0+20744+ffd02a6d.ppc64le"
                    ],
                    "SourcePackageList": [
                        "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.src",
                        "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.src",
                        "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.src",
                        "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.src",
                        "ipa-4.9.2-10.module+el8.4.0+20744+ffd02a6d.src"
                    ]
                },
                {
                    "Name": "idm",
                    "StreamID": 87242,
                    "Stream": "DL1",
                    "Version": "8040020231123154610",
                    "Context": "5b01ab7e",
                    "PackageList": [
                        "python3-pyusb-1.0.0-9.module+el8.1.0+4098+f286395e.noarch",
                        "python3-custodia-0.6.0-3.module+el8.1.0+4098+f286395e.noarch",
                        "python3-jwcrypto-0.5.0-1.module+el8.1.0+4098+f286395e.noarch",
                        "custodia-0.6.0-3.module+el8.1.0+4098+f286395e.noarch",
                        "python3-qrcode-core-5.1-12.module+el8.1.0+4098+f286395e.noarch",
                        "python3-qrcode-5.1-12.module+el8.1.0+4098+f286395e.noarch",
                        "python3-yubico-1.3.2-9.module+el8.1.0+4098+f286395e.noarch",
                        "python3-kdcproxy-0.4-5.module+el8.2.0+4691+a05b2456.noarch",
                        "ipa-healthcheck-0.7-3.module+el8.4.0+9007+5084bdd8.noarch",
                        "ipa-healthcheck-core-0.7-3.module+el8.4.0+9007+5084bdd8.noarch",
                        "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.x86_64",
                        "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.x86_64",
                        "softhsm-devel-2.6.0-5.module+el8.4.0+10227+076cd560.x86_64",
                        "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.x86_64",
                        "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.x86_64",
                        "ipa-client-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
                        "ipa-client-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-client-epn-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
                        "ipa-client-samba-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
                        "ipa-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-python-compat-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-selinux-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-server-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
                        "ipa-server-common-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-server-dns-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "ipa-server-trust-ad-4.9.2-10.module+el8.4.0+20744+ffd02a6d.x86_64",
                        "python3-ipaclient-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "python3-ipalib-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "python3-ipaserver-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch",
                        "python3-ipatests-4.9.2-10.module+el8.4.0+20744+ffd02a6d.noarch"
                    ],
                    "SourcePackageList": [
                        "python-qrcode-5.1-12.module+el8.1.0+4098+f286395e.src",
                        "custodia-0.6.0-3.module+el8.1.0+4098+f286395e.src",
                        "python-jwcrypto-0.5.0-1.module+el8.1.0+4098+f286395e.src",
                        "python-kdcproxy-0.4-5.module+el8.2.0+4691+a05b2456.src",
                        "python-yubico-1.3.2-9.module+el8.1.0+4098+f286395e.src",
                        "pyusb-1.0.0-9.module+el8.1.0+4098+f286395e.src",
                        "opendnssec-2.1.7-1.module+el8.4.0+9007+5084bdd8.src",
                        "bind-dyndb-ldap-11.6-2.module+el8.4.0+9328+4ec4e316.src",
                        "ipa-healthcheck-0.7-3.module+el8.4.0+9007+5084bdd8.src",
                        "softhsm-2.6.0-5.module+el8.4.0+10227+076cd560.src",
                        "slapi-nis-0.56.6-2.1.module+el8.4.0+12579+0538eac8.src",
                        "ipa-4.9.2-10.module+el8.4.0+20744+ffd02a6d.src"
                    ]
                }
            ],
 ```
it is reproducible with 
```go
	req := vmaas.ErrataRequest{ErrataList: []string{"RHSA-2024:0138"}}
	errata, _ := api.Errata(&req)
	resp, _ := json.Marshal(errata)
```
I can share with you a vmaas sqlite dump from stage if you want it for debugging.

File: vmaas/errata.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (psegedy) [reply vmaas/errata.go:15]: yeah the request allows either string or list of strings but the response spec specify it only as a string, the question is if the real response is the same as defined by the spec

[the remark this answers is not in the document]
Comment (psegedy) [reply vmaas/errata.go:15]: aaaand the reality is that the response is always []string, so we can get rid of the Marshal function and use []string in the response struct
```
~ curl https://console.redhat.com/api/vmaas/v3/errata -d '{"errata_list": ["foo"], "type": "security", "severity": "Important"}' -H "Content-type: application/json" | jq
{
  "errata_list": {},
  "last_change": "2024-10-31T14:46:46.626394+00:00",
  "type": [
    "security"
  ],
  "severity": [
    "Important"
  ],
  "page": 1,
  "page_size": 0,
  "pages": 0
}
```

**Review by psegedy — APPROVED** [review]

**Pull request merged**